### ISSUE TITLE
Support for Apple Silicon

### ISSUE THREAD

**Issue by danbricedatascience** [microsoft/LightGBM#3606]
When does LightGBM is (roughly) expected to support Apple Silicon ?

**Comment by StrikerRUS** [comment]
What do you mean by "support"?

I believe LightGBM can be compiled on Apple Silicon. If compilation fails, could you kindly provide any logs from your machine?

**Comment by danbricedatascience** [comment]
Hi, by support I mean not only compiling it but enable using every ML capabilities, so the 8 cores GPU, the two neural accelerators and the 16 cores Neural Engine for model training, so not using Cuda but a totally different approach. [Tensorflow already has an alpha version available](https://blog.tensorflow.org/2020/11/accelerating-tensorflow-performance-on-mac.html) for testing but seems far to be working perfectly in any situation for the moment, and Apple helps google on this task so it doesn't seem trivial.

Anyway recompiling it for CPU is a first step. Not sure if can be done as for instance numpy cannot be recompiled like this, it totally fails. Anyway I will try.

**Comment by StrikerRUS** [comment]
Thanks a lot for your fast response! To be honest, I don't think that this kind of support (Mac-optimized version actually) will be implemented in the near future. At least, not by our small maintaining team without any help from the outside.

Our current GPU implementation is far from perfect in terms of GPU utilization https://github.com/microsoft/LightGBM/issues/768#issuecomment-320573774 and new CUDA implementation kindly contributed by IBM folks has [some bugs](https://github.com/microsoft/LightGBM/pull/3160#issuecomment-724301806) that don't allow us even to [announce it](https://github.com/microsoft/LightGBM/pull/3428#issuecomment-721307376). So, I think this "general" GPU issues have more priority. Also, please note that Mac-optimized TensorFlow version are developing by Apple and Tensorflow-team is planing just adding that version as community-supported: 
> In the near future, we’ll be making updates like this even easier for users to get these performance numbers by integrating the forked version into the [TensorFlow master branch](https://github.com/tensorflow/tensorflow#community-supported-builds).

**Comment by StrikerRUS** [comment]
Linking https://github.com/dmlc/xgboost/issues/6408 here.

**Comment by danbricedatascience** [comment]
@StrikerRUS Thanks for your answer. Anyway if it can at least work in standard CPU mode it could already be great. I will try to compile when I will have time and let you know.

**Comment by guolinke** [comment]
@danbricedatascience I think the "ML engine" is better for neural networks, not for decision trees.

**Comment by danbricedatascience** [comment]
@guolinke yes maybe it's a pure linear algebra unit, like a TPU, but this is the same for GPUs, it's pure SIMD units mostly used for vector computing in ML. I don't know how LightGBM or the others Gradient Boosting are using GPUs but they are, so at least if it cannot use the M1 Neural Engine, it can maybe use its GPU  ? This 2.2 TFlops unit that has nothing to do with Intel integrated GPUs, check benchmarks.

Apple says [here](https://machinelearning.apple.com/updates/ml-compute-training-on-mac):

"Until now, TensorFlow has only utilized the CPU for training on Mac. The new tensorflow_macos fork of TensorFlow 2.4 leverages ML Compute to enable machine learning libraries to take full advantage of not only the CPU, but also the GPU in both M1- and Intel-powered Macs for dramatically faster training performance. This starts by applying higher-level optimizations such as fusing layers, selecting the appropriate device type and compiling and executing the graph as primitives that are accelerated by BNNS on the CPU and Metal Performance Shaders on the GPU."

**Comment by StrikerRUS** [comment]
Closed in favor of being in #2302. We decided to keep all feature requests in one place.

Welcome to contribute this feature! Please re-open this issue (or post a comment if you are not a topic starter) if you are actively working on implementing this feature.
